server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java

```diff
@@ -19,20 +19,120 @@
 
-    public SetMailboxesResponse process(SetMailboxesRequest request) {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SetMailboxesCreationProcessor.class);
+
+    private final MailboxManager mailboxManager;
+    private final CollectionHierarchySorter<Map.Entry<MailboxCreationId, MailboxRequest>, String> collectionHierarchySorter;
+    private final MailboxUtils<Id> mailboxUtils;
+
+    @Inject
+    @VisibleForTesting
+    SetMailboxesCreationProcessor(MailboxManager mailboxManager, MailboxUtils<Id> mailboxUtils) {
+        this.mailboxManager = mailboxManager;
+        this.collectionHierarchySorter =
+            new CollectionHierarchySorter<>(
+                x -> x.getKey().getCreationId(),
+                x -> x.getValue().getParentId());
+        this.mailboxUtils = mailboxUtils;
+    }
+
+    public SetMailboxesResponse process(SetMailboxesRequest request, MailboxSession mailboxSession) {
         SetMailboxesResponse.Builder builder = SetMailboxesResponse.builder();
-        request.getCreate().entrySet().stream()
-            .forEach(entry -> builder.creation(
-                entry.getKey(),
-                Mailbox.builder().name(entry.getValue().getName()).id("serverId").build()));
+        try {
+            Map<MailboxCreationId, String> creationIdsToCreatedMailboxId = new HashMap<>();
+            collectionHierarchySorter.sortFromRootToLeaf(request.getCreate().entrySet())
+                .forEach(entry -> 
+                    createMailbox(entry.getKey(), entry.getValue(), mailboxSession, creationIdsToCreatedMailboxId, builder));
+        } catch (CycleDetectedException e) {
+            request.getCreate().entrySet()
+                .forEach(entry ->
+                    builder.notCreated(entry.getKey(),
+                            SetError.builder()
+                            .type("invalidArguments")
+                            .description("The created mailboxes introduce a cycle.")
+                            .build()));
+        }
         return builder.build();
     }
 
+    private void createMailbox(MailboxCreationId mailboxCreationId, MailboxRequest mailboxRequest, MailboxSession mailboxSession,
+            Map<MailboxCreationId, String> creationIdsToCreatedMailboxId, SetMailboxesResponse.Builder builder) {
+        try {
+            MailboxPath mailboxPath = getMailboxPath(mailboxRequest, creationIdsToCreatedMailboxId, mailboxSession);
+            mailboxManager.createMailbox(mailboxPath, mailboxSession);
+            Optional<Mailbox> mailbox = mailboxUtils.mailboxFromMailboxPath(mailboxPath, mailboxSession);
+            if (mailbox.isPresent()) {
+                builder.creation(mailboxCreationId, mailbox.get());
+                creationIdsToCreatedMailboxId.put(mailboxCreationId, mailbox.get().getId());
+            } else {
+                builder.notCreated(mailboxCreationId, SetError.builder()
```

Comment:
As the `MailboxManager.createMailbox` returns nothing, i have to get the mailbox after creation in order to send it to the response.
The `MailboxManager.createMailbox` method is not sexy; i don't really want to make it return something:
```java
    public void createMailbox(MailboxPath mailboxPath, final MailboxSession mailboxSession)
            throws MailboxException {
        mailboxSession.getLog().debug("createMailbox " + mailboxPath);
        final int length = mailboxPath.getName().length();
        if (length == 0) {
            mailboxSession.getLog().warn("Ignoring mailbox with empty name");
        } else {
            if (mailboxPath.getName().charAt(length - 1) == getDelimiter())
                mailboxPath.setName(mailboxPath.getName().substring(0, length - 1));
            if (mailboxExists(mailboxPath, mailboxSession))
                throw new MailboxExistsException(mailboxPath.toString());
            // Create parents first
            // If any creation fails then the mailbox will not be created
            // TODO: transaction
            for (final MailboxPath mailbox : mailboxPath.getHierarchyLevels(getDelimiter()))

                locker.executeWithLock(mailboxSession, mailbox, new LockAwareExecution<Void>() {

                    public Void execute() throws MailboxException {
                        if (!mailboxExists(mailbox, mailboxSession)) {
                            final org.apache.james.mailbox.store.mail.model.Mailbox<Id> m = doCreateMailbox(mailbox, mailboxSession);
                            final MailboxMapper<Id> mapper = mailboxSessionMapperFactory.getMailboxMapper(mailboxSession);
                            mapper.execute(new TransactionalMapper.VoidTransaction() {

                                public void runVoid() throws MailboxException {
                                    mapper.save(m);
                                }

                            });

                            // notify listeners
                            dispatcher.mailboxAdded(mailboxSession, m);
                        }
                        return null;

                    }
                }, true);

        }
    }
```